Browser, LibWeb, LibGfx, AK: Added size preview widget in inspector

MichielVrins opened the pull request:
**Browser, LibWeb:**
This Adds an element size preview widget to the inspector widgetin a new tab. This functions similar to chrome and firefox and shows the margin, border, padding, and content size of the selected element in the inspector.

The rectangles automatically resize if long sizes are displayed within them but are capped at 4 decimals.

The colors for the size preview widget are taken from the chrome browser.

**LibGfx:**
For this TextAlignment::TopCenter and TextAlignment::BottomCenter had to be added to Painter::draw_text.

**AK:**
Added to_double() in JsonValue. 

![image](https://user-images.githubusercontent.com/10757347/155864043-30f537d7-2bab-4171-9163-8c9caff05b02.png)

drunderscore commented on Userland/Applications/Browser/ElementSizePreviewWidget.h:
```diff
@@ -0,0 +1,31 @@
+private:
+    virtual void paint_event(GUI::PaintEvent&) override;
+    Web::Layout::BoxModelMetrics m_node_box_sizing;
+    float m_node_content_height;
+    float m_node_content_width;
```
Let's initialize these primitive members!

MichielVrins replied:
Fixed and force pushed.

drunderscore reviewed (commented):
Just one small change, but this looks incredibly cool!

GMTA commented:
Very cool! Small visual nitpick: in your screenshot, those top 4/0/15 numbers are a bit too close to the horizontal line above them when compared to the padding of the other numbers.

Hendiadyoin1 commented on Userland/Applications/Browser/InspectorWidget.cpp:
```diff
@@ -151,6 +160,36 @@ void InspectorWidget::load_style_json(String specified_values_json, String compu
+    auto json_value = json_or_error.release_value();
+    const auto& json_object = json_value.as_object();
+
+    m_node_box_sizing.margin.top = static_cast<float>(json_object.get("margin_top").to_double());
+    m_node_box_sizing.margin.right = static_cast<float>(json_object.get("margin_right").to_double());
+    m_node_box_sizing.margin.bottom = static_cast<float>(json_object.get("margin_bottom").to_double());
+    m_node_box_sizing.margin.left = static_cast<float>(json_object.get("margin_left").to_double());
+    m_node_box_sizing.padding.top = static_cast<float>(json_object.get("padding_top").to_double());
+    m_node_box_sizing.padding.right = static_cast<float>(json_object.get("padding_right").to_double());
+    m_node_box_sizing.padding.bottom = static_cast<float>(json_object.get("padding_bottom").to_double());
+    m_node_box_sizing.padding.left = static_cast<float>(json_object.get("padding_left").to_double());
+    m_node_box_sizing.border.top = static_cast<float>(json_object.get("border_top").to_double());
+    m_node_box_sizing.border.right = static_cast<float>(json_object.get("border_right").to_double());
+    m_node_box_sizing.border.bottom = static_cast<float>(json_object.get("border_bottom").to_double());
+    m_node_box_sizing.border.left = static_cast<float>(json_object.get("border_left").to_double());
```
couldn't you just have added a to_float?

MichielVrins replied:
Fixed and force pushed.

MichielVrins commented:
> Very cool! Small visual nitpick: in your screenshot, those top 4/0/15 numbers are a bit too close to the horizontal line above them when compared to the margin of the other numbers.

This is now fixed. As it happens to be, the glyphs for numbers are substantially larger vertically than letters. To account for this i just increased the padding for this text, because to my understanding,  its not possible to center both numbers and letters perfectly vertically when they're on the same text line. 

![image](https://user-images.githubusercontent.com/10757347/156923163-f48b14ff-c841-4457-8e61-fa7b3be4dfe0.png)

> couldn't you just have added a to_float?

I added an extra commit here to add support for floats in Json and used that in the serializer for the inspector instead of casting floats to doubles.

Hendiadyoin1 commented on AK/JsonObjectSerializer.h:
```diff
@@ -129,6 +129,16 @@ class JsonObjectSerializer {
         return {};
     }
 
+    ErrorOr<void> add(StringView key, float value)
+    {
+        TRY(begin_item(key));
+        if constexpr (IsLegacyBuilder<Builder>)
+            TRY(m_builder.try_appendff("{}", value));
+        else
+            TRY(m_builder.appendff("{}", value));
+        return {};
+    }
+
```
Please put the `float` code always next to the double code and into the KERNEL guard

MichielVrins replied:
After further investigation I fully agree. I have removed this interface and kept only the `to_float()` and `ErrorOr<void> add(StringView key, float value)` and moved them into the kernel guards.

Hendiadyoin1 commented on AK/JsonValue.h:
```diff
@@ -102,6 +102,13 @@ class JsonValue {
     u32 to_u32(u32 default_value = 0) const { return to_number<u32>(default_value); }
     u64 to_u64(u64 default_value = 0) const { return to_number<u64>(default_value); }
 
+#if !defined(KERNEL)
```
`#ifndef`

AtkinsSJ replied:
IDK, the codebase uses `defined()` about as much as `ifdef/ifndef`. I don't think it's a big enough deal to require changing it.

Hendiadyoin1 reviewed (commented):
Thinking about it I don't like adding a float interface to JSON, because the Parser will still only produce doubles, so just using `to_number<float>` would practically achieve the same...

Here are some Nitpicks anyway

linusg reviewed (changes requested):
https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy

> Write the commit message subject line in the imperative mood ("Foo: Change the way dates work", not "Foo: Changed the way dates work").

Please also use `Foo+Bar: Message` instead of `Foo, Bar: Message` for multiple categories.

linusg commented:
You didn't actually change anything?

MichielVrins commented:
> You didn't actually change anything?

Ok, this time they are actually changed (misinterpreted the rebase command, was trying to change them with pick instead of reword)
